### ISSUE TITLE
error from find

### ISSUE THREAD

**Issue by VaughanR0** [Metadrop/drupal-fix-permissions-script#2]
Running this command on a linux server
./drupal_fix_permissions.sh -n -u=webmaster -g=apache -p=/usr/share/drupal7
returns this error from find
find: paths must precede expression: /usr/share/drupal7/sites/xxx/files

NB /usr/share/drupal7/sites is a symbolic link to /etc/drupal7

These tweaks seemed to fix the issue:

    # Real action.
    find "$1" \( -path "./sites/*/$file_folder_name" -prune \) -o \( -path "./sites/*/$private_folder_name" -prune \) -o \( -type $2 ! -perm $3 -print0 \) | xargs -r -0 -L4 chmod $3

...

    # Simulate.
    num=$(find "$1" \( -path "./sites/*/$file_folder_name" -prune \) -o \( -path "./sites/*/$private_folder_name" -prune \) -o \( -type $2 ! -perm $3 -print \) | wc -l)

and 

    # Simulate verbosely.
    printf "\n    Code files and directories that would have their permissions fixed: "
    # Use a variable to indent output.
    items=$(find "$1" \( -path "./sites/*/$file_folder_name" -prune \) -o \( -path "./sites/*/$private_folder_name" -prune \) -o \( -type $2 ! -perm $3 -print \))

**Comment by rsanzante** [comment]
I'm not able to reproduce the problem.  I've tried setting up a Drupal locally, moving sites folder outside of the docroot folder and symlinking it. However, the script still works ok.

I've tried with a Drupal 10. The script was developed for Drupal 8/9/10 and I'm not sure if it works as is for a Drupal 7 installation, I need to perform some tests on that.

Additionally, the proposed changes break the functionality that makes sure only files that need to be updated are processed. See https://github.com/Metadrop/drupal-fix-permissions-script/tree/main#performance. For example, the `\( ! -user $drupal_user -o ! -group $httpd_group \)` is lost. Also, the `$detected_vendor_path` is missing, so the `vendor` folder is not processed.

**Comment by VaughanR0** [comment]
Hi Ricardo, I am sorry to take your time, but I thought you might like to know that your script did not work on my machine.  The changes I suggested only applied to the use of find in three places in the fix_code_permission_helper function.
We are probably using a different version of find.  Mine is GNU find version 4.5.11 with 
Features enabled: D_TYPE O_NOFOLLOW(enabled) LEAF_OPTIMISATION FTS(FTS_CWDFD) CBO(level=2)

**Comment by rsanzante** [comment]
hi @VaughanR0, I appreciate your time creating the bug report, I would like this script to work on as many systems as possible. 

On the environments I've tested `find` version is 4.8.0 adn same features: `D_TYPE O_NOFOLLOW(enabled) LEAF_OPTIMISATION FTS(FTS_CWDFD) CBO(level=2) `

The problem with the proposed changes are:
  - Missing $vendor in the target folder list
  - Find targets all files instead of exclude the ones that don't need their permissions fixed

Also, as said before, I can't reproduce the bug. So I guess either:
  - The bug arises on Drupal 7 folder structure
  - There's something in your system that I'm not aware of that triggers the bug

I think I would need to clarify this to push this forward.

**Comment by rsanzante** [comment]
I've tested with a fresh Drupal 7 download and it worked. 

I've used this to get the code:

```
composer create-project drupal-composer/drupal-project:7.x-dev freshd7_dir
```

The ran the script:

```
sudo ~/development/metadrop/drupal-fix-permissions-github/drupal_fix_permissions.sh -u=ricardo
Script to fix permissions in a Drupal installation
Using default Drupal path '.'
Using default HTTPD group 'www-data'
Running configuration:
Path: /tmp/composer/freshd7_dir/web
Owner user: ricardo
Owner group: www-data
Code dirs perms: u=rwx,g=rx,o=
Code files perms: u=rw,g=r,o=
Separated vendor folder detected: ../vendor
Content dirs perms: u=rwx,g=rwx,o=
Content files perms: ug=rw,o=
File folder name: files
Private files folder name: private


All checks passed, go!
Processing Drupal installed on '/tmp/composer/freshd7_dir/web'
Fixing ownership of files and directories
Fixing permissions on code files and directories
  Setting permissions on code directories to u=rwx,g=rx,o= under 'web'
  Setting permissions on code files to u=rw,g=r,o= under 'web'
  Setting permissions on vendor code directories to u=rwx,g=rx,o= under '../vendor'
  Removing all permissions on vendor code files to other users (o=) under '../vendor'
Fixing permissions on content files and directories under 'sites' folder
  Checking folder default
    Setting permissions on content directories to u=rwx,g=rwx,o= under 'files'
    Setting permissions on content files to ug=rw,o= under 'files'
  Checking folder all
Processing additional content folders

Permissions and ownership fixed!                    
```

Can you run the script with bash -x and paste the log?

**Comment by VaughanR0** [comment]
Hi Ricardo,

I am sorry, but I only have a production system and so I can’t do development on it.

Best regards,
Vaughan

**Comment by rsanzante** [comment]
Closed due to inactivity.

**Comment by fenstrat** [comment]
I'm also getting the error `find: paths must precede expression:`. On a Drupal 10 platform.

Getting it when this following is running; `Setting permissions on code directories to `, and `Setting permissions on code files to `.

I only hit the issue when there is 2 or more site. I.e. if there is `sites/default` and `sites/site1.com` then there's no find error. However if there is a 3rd site `sites/site2.com` then the find error occurs.

Also confirming the tweaks @VaughanR0 proposed above, i.e. removing `"$1"` from the -path param (i.e. `-path "$1"/sites` > `-path "./sites`) seem to fix the issue.

I don't have any symlinks. Same version of `find` as @VaughanR0. I've not got any more time to dig into this right now, but wanted to report some steps to replicate it.

**Comment by rsanzante** [comment]
Issue reopened per due to information provided

**Comment by rsanzante** [comment]
Could it be related to the find version? I'm seeing this error using find  4.5.11 but not with find 4.8.0.

For those affected by this bug, can you post the find version you are using?

**Comment by rsanzante** [comment]
Forget previous comment, I think the problem is an asterisk that is not escaped.

**Comment by rsanzante** [comment]
Fixed in https://github.com/Metadrop/drupal-fix-permissions-script/commit/3588e001da55358dce9bc9e15d810d8b1b1aa782

Please, can you download the dev version and try it?

You can download the script directly from here: https://raw.githubusercontent.com/Metadrop/drupal-fix-permissions-script/3588e001da55358dce9bc9e15d810d8b1b1aa782/drupal_fix_permissions.sh

**Comment by fenstrat** [comment]
@rsanzante fantastic, confirming that that escaped `*` that was fixed in 3588e001da55358dce9bc9e15d810d8b1b1aa782 got rid of the `find: paths must precede expression:` error from find. Thanks!

**Comment by rsanzante** [comment]
Thanks for que quick response, I'm closing this and releasing a new version.
